{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import cryptofxns\n",
    "import os\n",
    "from os import listdir\n",
    "from matplotlib import pyplot as plt\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "hap_filter_size = 0\n",
    "HAPFILTINT = int(hap_filter_size/1000.0)\n",
    "Method = '-min'\n",
    "fileendname = Method +'-Geno-k' + str(int(hap_filter_size)) + '.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "genodf = cryptofxns.geno_df;genodf.shape\n",
    "centromere = cryptofxns.load_centrom_locs()\n",
    "clens = cryptofxns.chromosome_lengths\n",
    "tc = cryptofxns.thecol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "94"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chrlist = np.unique(genodf.Chrom).tolist()\n",
    "segs_parents = genodf.columns.tolist()[tc:];\n",
    "segs_drop = cryptofxns.geno_remove;\n",
    "segs = [s for s in segs_parents if s not in segs_drop+ ['SS-A837','XL280','XL280a']]\n",
    "len(segs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(86767, 98)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genodf = genodf[genodf.columns.tolist()[:tc]+segs].dropna(how='any',axis=0);\n",
    "genodf.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "blakes_bams = pd.read_csv('../FILES/blakes_bams_qtl.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "107"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(blakes_bams.columns.tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "108"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(segs_parents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "FILECHRPATHS = cryptofxns.FILECHRPATHS;\n",
    "FGTFILES = [f[3] for f in FILECHRPATHS];"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'../FILES/CHR1/FGT/'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "FGTFILES[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'-min-Geno-k0.csv'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fileendname"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#SS-B410-min-haplotype-k0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "fgtdfs = [];\n",
    "for ch, chrom in enumerate(chrlist):\n",
    "    dfs = [];\n",
    "    for seg in segs:#for tetrad in tetrads:\n",
    "        path = FGTFILES[ch] + seg + '/'+ seg + fileendname#-min-haplotype-k6000.csv'\n",
    "        if os.path.exists(path):\n",
    "            dfs.append(pd.read_csv(path));\n",
    "    fgtdfs.append(dfs);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "ngenodf = pd.DataFrame(index=genodf.index.tolist(),columns=genodf.columns.tolist())\n",
    "ngenodf['Chrom'] = genodf.Chrom;\n",
    "ngenodf['Alt'] = genodf.Alt;\n",
    "ngenodf['Ref'] = genodf.Ref;\n",
    "ngenodf['Pos'] = genodf.Pos;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Alt</th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Pos</th>\n",
       "      <th>Ref</th>\n",
       "      <th>SS-B307</th>\n",
       "      <th>SS-B308</th>\n",
       "      <th>SS-B310</th>\n",
       "      <th>SS-B311</th>\n",
       "      <th>SS-B312</th>\n",
       "      <th>SS-B313</th>\n",
       "      <th>...</th>\n",
       "      <th>SS-C029_cor</th>\n",
       "      <th>SS-C031_cor</th>\n",
       "      <th>SS-C033</th>\n",
       "      <th>SS-C036</th>\n",
       "      <th>SS-C039</th>\n",
       "      <th>SS-C242</th>\n",
       "      <th>SS-C271</th>\n",
       "      <th>SS-C272</th>\n",
       "      <th>SS-C290</th>\n",
       "      <th>SS-C291</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>5016.0</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>5299.0</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>5464.0</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>6120.0</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>6166.0</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 98 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Alt  Chrom     Pos Ref SS-B307 SS-B308 SS-B310 SS-B311 SS-B312 SS-B313  \\\n",
       "0   A  Chr01  5016.0   C     NaN     NaN     NaN     NaN     NaN     NaN   \n",
       "1   C  Chr01  5299.0   T     NaN     NaN     NaN     NaN     NaN     NaN   \n",
       "2   C  Chr01  5464.0   T     NaN     NaN     NaN     NaN     NaN     NaN   \n",
       "3   C  Chr01  6120.0   T     NaN     NaN     NaN     NaN     NaN     NaN   \n",
       "4   A  Chr01  6166.0   G     NaN     NaN     NaN     NaN     NaN     NaN   \n",
       "\n",
       "    ...   SS-C029_cor SS-C031_cor SS-C033 SS-C036 SS-C039 SS-C242 SS-C271  \\\n",
       "0   ...           NaN         NaN     NaN     NaN     NaN     NaN     NaN   \n",
       "1   ...           NaN         NaN     NaN     NaN     NaN     NaN     NaN   \n",
       "2   ...           NaN         NaN     NaN     NaN     NaN     NaN     NaN   \n",
       "3   ...           NaN         NaN     NaN     NaN     NaN     NaN     NaN   \n",
       "4   ...           NaN         NaN     NaN     NaN     NaN     NaN     NaN   \n",
       "\n",
       "  SS-C272 SS-C290 SS-C291  \n",
       "0     NaN     NaN     NaN  \n",
       "1     NaN     NaN     NaN  \n",
       "2     NaN     NaN     NaN  \n",
       "3     NaN     NaN     NaN  \n",
       "4     NaN     NaN     NaN  \n",
       "\n",
       "[5 rows x 98 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ngenodf.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fill in the dataframe using old index but filtered genotype values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fgtdfs[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "for ch,chrom in enumerate(chrlist):\n",
    "    inxs = genodf[genodf['Chrom'] == chrom].index.tolist() ## This only works if you take SNPs with CR of 100%\n",
    "    for seg in fgtdfs[ch]:\n",
    "        seg_name = seg.columns.tolist()[1]\n",
    "        #inxs = [ngenodf[(ngenodf['Chrom']==chrom)&(ngenodf['Pos']==pos)].index[0] for pos in seg['Pos'].tolist()]; ## THIS TAKES TO LONG! \n",
    "        #inxs = genodf[genodf['Chrom'] == chrom][seg_name].dropna().index.tolist() ## Take index from old genotype dataframe\n",
    "        ngenodf.ix[inxs,seg_name] = seg[seg_name].tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Drop the parental strains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "ngenodf.dropna(axis=1,how='all',inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## View results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(86767, 4)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ngenodf.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Alt</th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Pos</th>\n",
       "      <th>Ref</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>5016.0</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>5299.0</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>5464.0</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>6120.0</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>6166.0</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Alt  Chrom     Pos Ref\n",
       "0   A  Chr01  5016.0   C\n",
       "1   C  Chr01  5299.0   T\n",
       "2   C  Chr01  5464.0   T\n",
       "3   C  Chr01  6120.0   T\n",
       "4   A  Chr01  6166.0   G"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ngenodf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Alt</th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Pos</th>\n",
       "      <th>Ref</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>96332</th>\n",
       "      <td>C</td>\n",
       "      <td>Chr09</td>\n",
       "      <td>1176293.0</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96333</th>\n",
       "      <td>C</td>\n",
       "      <td>Chr09</td>\n",
       "      <td>1176398.0</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96334</th>\n",
       "      <td>C</td>\n",
       "      <td>Chr09</td>\n",
       "      <td>1177613.0</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96335</th>\n",
       "      <td>T</td>\n",
       "      <td>Chr09</td>\n",
       "      <td>1178017.0</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96336</th>\n",
       "      <td>A</td>\n",
       "      <td>Chr09</td>\n",
       "      <td>1178428.0</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Alt  Chrom        Pos Ref\n",
       "96332   C  Chr09  1176293.0   A\n",
       "96333   C  Chr09  1176398.0   A\n",
       "96334   C  Chr09  1177613.0   G\n",
       "96335   T  Chr09  1178017.0   A\n",
       "96336   A  Chr09  1178428.0   G"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ngenodf.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert 14 == len(np.unique(ngenodf.Chrom.tolist()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis of Intermarker Intervals\n",
    "## Definition: Intermarker Interval\n",
    "#### Intermarker: For consecutive markers [v1, w1] and [v2, w2], the intermarker interval (IMI), defined as [x,y] â‰¡ [w1 , v2 ] âŠ‚ Z.  The nominal size of the IMI is then y âˆ’ x + 1.  Note that by this definition, markers separated by 1 bp or less produce an IMI size of 0."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " Total Number IMIs: 86753\n",
      "        IMIs < 2kb: 86278\n",
      "      % IMIs > 2kb: 0.548\n",
      "   Median IMI size: 87.0 bp\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAagAAAEYCAYAAAAJeGK1AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4wLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvpW3flQAAHTpJREFUeJzt3X2UJXV95/H3J4xPMUFABoKAGYwT\nlbhRcQLjuroqBgbMCiYhYh4YlbMTPZgTzcM6rtlANJ5onjQkEcPGWYYsEQ3GMEZwnEV8jCiDgoiA\nMyLKCMLo8JQYMch3/6hfw6W5PdO3u2/fgnm/zrnn1v3Wr6q+t253f7uqfrd+qSokSeqbH5p0ApIk\nDWOBkiT1kgVKktRLFihJUi9ZoCRJvWSBkiT1kgVKktRLFihJUi9ZoCRJvbRk0gn00b777lvLli2b\ndBqS9JB02WWXfbuqlu6qnQVqiGXLlrF58+ZJpyFJD0lJvj6bdp7ikyT1kgVKktRLFihJUi9ZoCRJ\nvWSBkiT1kgVKktRLFihJUi9ZoCRJvWSBkiT1kneSWGDL1n5ootu//q0vmuj2JWmheAQlSeolC5Qk\nqZcmXqCS7JXkvCTXJLk6ybOS7JNkU5It7Xnv1jZJTk+yNckXkxw2sJ7Vrf2WJKsH4s9McmVb5vQk\nmcT7lCSNZuIFCvgL4MNV9WTgacDVwFrgoqpaDlzUXgMcAyxvjzXAGQBJ9gFOBY4ADgdOnSpqrc2a\ngeVWLcJ7kiTN00QLVJI9gecC7waoqu9X1W3AccD61mw9cHybPg44uzqXAHslOQA4GthUVTuq6lZg\nE7Cqzduzqj5TVQWcPbAuSVKPTfoI6gnAduD/JPlCkr9N8mhg/6q6CaA979faHwjcMLD8thbbWXzb\nkPgDJFmTZHOSzdu3b5//O5MkzcukC9QS4DDgjKp6BvBv3Hc6b5hh149qDvEHBqvOrKoVVbVi6dJd\nDvQoSRqzSReobcC2qvpse30eXcG6uZ2eoz3fMtD+4IHlDwJu3EX8oCFxSVLPTbRAVdW3gBuSPKmF\njgS+DGwApnrirQbOb9MbgJNab76VwO3tFOBG4Kgke7fOEUcBG9u8O5OsbL33ThpYlySpx/pwJ4nf\nAM5J8nDgOuAVdIXzfUlOBr4BnNDaXgAcC2wFvtvaUlU7krwZuLS1e1NV7WjTrwbOAh4FXNgekqSe\nm3iBqqrLgRVDZh05pG0Bp8ywnnXAuiHxzcBT55mmJGmRTfoalCRJQ1mgJEm9ZIGSJPWSBUqS1EsW\nKElSL1mgJEm9ZIGSJPWSBUqS1EsWKElSL1mgJEm9ZIGSJPWSBUqS1EsWKElSL1mgJEm9ZIGSJPWS\nBUqS1EsWKElSL1mgJEm9ZIGSJPWSBUqS1EsWKElSL1mgJEm9ZIGSJPWSBUqS1EsWKElSL028QCW5\nPsmVSS5PsrnF9kmyKcmW9rx3iyfJ6Um2JvliksMG1rO6td+SZPVA/Jlt/Vvbsln8dylJGtXEC1Tz\n/Kp6elWtaK/XAhdV1XLgovYa4BhgeXusAc6ArqABpwJHAIcDp04VtdZmzcByq8b/diRJ89WXAjXd\nccD6Nr0eOH4gfnZ1LgH2SnIAcDSwqap2VNWtwCZgVZu3Z1V9pqoKOHtgXZKkHutDgSrgI0kuS7Km\nxfavqpsA2vN+LX4gcMPAsttabGfxbUPiD5BkTZLNSTZv3759nm9JkjRfSyadAPDsqroxyX7ApiTX\n7KTtsOtHNYf4A4NVZwJnAqxYsWJoG0nS4pn4EVRV3diebwE+QHcN6eZ2eo72fEtrvg04eGDxg4Ab\ndxE/aEhcktRzEy1QSR6d5EenpoGjgC8BG4CpnnirgfPb9AbgpNabbyVwezsFuBE4KsnerXPEUcDG\nNu/OJCtb772TBtYlSeqxSZ/i2x/4QOv5vQT4+6r6cJJLgfclORn4BnBCa38BcCywFfgu8AqAqtqR\n5M3Apa3dm6pqR5t+NXAW8CjgwvaQJPXcRAtUVV0HPG1I/DvAkUPiBZwyw7rWAeuGxDcDT513spKk\nRTXxa1CSJA1jgZIk9ZIFSpLUSxYoSVIvWaAkSb1kgZIk9ZIFSpLUSxYoSVIvWaAkSb1kgZIk9ZIF\nSpLUSxYoSVIvWaAkSb1kgZIk9ZIFSpLUSxYoSVIvWaAkSb1kgZIk9ZIFSpLUSxYoSVIvWaAkSb1k\ngZIk9ZIFSpLUSxYoSVIvWaAkSb00UoFKsneSQ5M8fFr85Unen+ScJM8aNYkkeyT5QpJ/bq8PSfLZ\nJFuSvHdqe0ke0V5vbfOXDazjDS1+bZKjB+KrWmxrkrWj5iZJmoxRj6D+GPjM4HJJfgt4N/AS4GXA\nxUmeNuJ6fxO4euD124C3V9Vy4Fbg5BY/Gbi1qp4IvL21I8mhwInATwGrgHe2orcH8NfAMcChwMta\nW0lSz41aoJ4LbKqq7wEkCfA7wFbgKcDzge8Dr5/tCpMcBLwI+NuBdb4AOK81WQ8c36aPa69p849s\n7Y8Dzq2qu6rqay2fw9tja1VdV1XfB85tbSVJPTdqgfox4GsDr3+6xf6iqq6tqo8DHwBGOc33DuB/\nAPe0148Fbququ9vrbcCBbfpA4AaANv/21v7e+LRlZoo/QJI1STYn2bx9+/YR0pckjcOoBWp6++cB\nBfy/gdg2uqK1S0l+Drilqi4bDA9pWruYN2r8gcGqM6tqRVWtWLp06U6yliQthiUjtr+e7rTZlOOB\nG6rqKwOxA+iuG83Gs4EXJzkWeCSwJ90R1V5JlrSjpIOAG1v7bcDBwLYkS4DHADsG4lMGl5kpLknq\nsVGPoP4eeE6S85KcTXdN6j3T2jwD2DKblVXVG6rqoKpaRtfJ4aNV9SvAxcAvtmargfPb9Ib2mjb/\no1VVLX5i6+V3CLAc+BxwKbC89Qp8eNvGhhHfsyRpAkY9gno7sIKux16ADwNvnpqZ5JnA04DT5pnX\n64Fzk/wh8AW6XoK0579LspXuyOlEgKq6Ksn7gC8DdwOnVNUPWk6vATYCewDrquqqeeYmSVoEIxWo\n1nvvF5Ls2b2sO6c1+TrdEdT1oyZSVR8DPtamr+P+pxIHt3/CDMu/BXjLkPgFwAWj5iNJmqxRv6j7\n3CSPr6o7hhQnqurbdEc2o34PSpKk+xn1GtTFwMt30eak1k6SpDkbtUAN67Y93R7M0JVbkqTZGsfN\nYn+a2XczlyRpqF12kkjy0Wmhlyd53pCme9DdpeEQ4P3zT02StDubTS++5w1MF7CsPaa7h66DxHnA\na+eZlyRpN7fLAlVVg3cuvwc4rareNNasJEm7vVG/qPt85vAdJ0mSRjXqF3U/Pq5EJEkaNOoRFEke\nQXeT2BXAXnSdI6arqjp5SFySpFkZqUAleQLwEbqeejv7TlRx3yi4kiSNbNQjqHcATwDOao9v0t2c\nVZKkBTVqgXoesLGqXjmGXCRJuteod5K4B7hyHIlIkjRo1AL1SbrhNCRJGqtRC9RvA89I4p0iJElj\nNeo1qLXAl4A/S3IKcAVwx5B2djOXJM3LqAXq5QPTP9Eew9jNXJI0L6MWqEPGkoUkSdOMequjr48r\nEUmSBo1jwEJJkuZt1FsdnTTbtlV19ujpSJLUGfUa1Fl0HSB2Jq2NBUqSNGejFqhXzBDfEzgM+GXg\nn4EN80lKkqRRO0ms39n8JGcAFwPvmk9SkiQtaCeJqvoc3dGTQ8JLkuZlHL34vgn89GwaJnlkks8l\nuSLJVUn+oMUPSfLZJFuSvDfJw1v8Ee311jZ/2cC63tDi1yY5eiC+qsW2Jlm7oO9UkjQ2C1qgkjwM\n+Fngzlkuchfwgqp6GvB0YFWSlcDbgLdX1XLgVu67K8XJwK1V9UTg7a0dSQ4FTgR+ClgFvDPJHkn2\nAP4aOAY4FHhZaytJ6rmF6ma+BHgc8FK6QvDO2ayvqgr41/byYe1RwAvoOlwArAdOA84AjmvTAOcB\nf5UkLX5uVd0FfC3JVuDw1m5rVV3X8j+3tf3ybPKTJE3OQnUznxr+vYB/AH53titsRzmXAU+kO9r5\nKnBbVU2N1LsNOLBNHwjcAFBVdye5HXhsi18ysNrBZW6YFj9ihjzWAGsAHv/4x882fUnSmCxUN/N7\ngNuAy6rqxlFWWFU/AJ6eZC/gA8BThjVrz5lh3kzxYacwh36Pq6rOBM4EWLFixa6+6yVJGrMF7WY+\nH1V1W5KPASuBvZIsaUdRBwFTRW8bcDCwLckS4DHAjoH4lMFlZopLknpsovfiS7K0HTmR5FHAC4Gr\n6b5L9Yut2Wrg/Da9ob2mzf9ou461ATix9fI7BFgOfA64FFjeegU+nK4jhV8ilqQHgVFP8QGQ5Bi6\nQvE0urtI3AF8ATi7qj48wqoOANa361A/BLyvqv45yZeBc5P8YVvvu1v7dwN/1zpB7KArOFTVVUne\nR9f54W7glHbqkCSvATYCewDrquqqubxnSdLiGrUXX4B1wEl0133uBr5DN3Dhk4CXJjm7qma6VnU/\nVfVF4BlD4tdxXy+8wfj3gBNmWNdbgLcMiV8AXDCbfCRJ/THqKb5T6I6cPg38V+CRVXUA8EjgucCn\ngJPacPCSJM3ZqAXqZOAa4Miq+mRV3QNQVfdU1aforiFdC/z3hU1TkrS7GbVAPQn4YFX9x7CZLf5B\n4Cfnm5gkafc2aoG6i65r987s2dpJkjRnoxaozcAvta7cD9DiL6Xr3i1J0pyN2s38rXRdtj/fxn76\nBHAzsD9dJ4lX0R1B/dFCJilJ2v2MeieJi5Ksprtx61rg9QOzQ3cX89VVdfHCpShJ2h2N/EXdqjon\nyQeB47n/F3WvAP6pqu5Y2BQlSbujWRWoJM+l6xzx4ar6j1aEzp7W5mF04zndVlWfXPhUJUm7k112\nkmgD/F0EvGim7uVwbxfzY4CLkjx54VKUJO2OZtOLbw3wfeD3ZtH2f7W2r5pPUpIkzaZAHQlcVFXf\n3lXDqvoO3dHWC+ebmCRp9zabAnUI3RAYs3UNsGxO2UiS1MymQI36Zd6ZRriVJGnWZlN8vgWM0unh\nyXRf3pUkac5mU6A+DbwwyQG7atjaHEU37IYkSXM2mwL1N8CjgPdNDc8+TJLHAO8FHgG8a2HSkyTt\nrnb5Rd2q+lSSvwF+Hbg6ybuAjwPfbE0OpBu88NeBHwPOqKp/GVO+kqTdxGxvdXQK8K/A64DfHzI/\nwD3AnwJvWJjUJEm7s1kVqDZy7u8mORN4OfAsuqMl6DpR/Auwvqq2jCNJSdLuZ9S7mW8B3jimXCRJ\nuteo33GSJGlRWKAkSb1kgZIk9ZIFSpLUSxMtUEkOTnJxkquTXJXkN1t8nySbkmxpz3u3eJKcnmRr\nki8mOWxgXatb+y1tWPqp+DOTXNmWOT2J9wmUpAeBSR9B3Q38dlU9BVgJnNIGSFxLN8THcrrhO9a2\n9scAy9tjDXAGdAUNOBU4AjgcOHWqqLU2awaWW7UI70uSNE8TLVBVdVNVfb5N30k3rMeBwHHA+tZs\nPXB8mz4OOLs6lwB7tfv/HQ1sqqodVXUrsIlu+PkDgD2r6jNVVXTD1E+tS5LUY5M+grpXkmXAM4DP\nAvtX1U3QFTFgv9bsQOCGgcW2tdjO4tuGxIdtf02SzUk2b9++fb5vR5I0T70oUEl+BHg/8NqqumNn\nTYfEZhp/amfxBwarzqyqFVW1YunSpbtKWZI0ZhMvUEkeRleczqmqf2zhm6eG92jPt7T4NuDggcUP\nAm7cRfygIXFJUs9NuhdfgHcDV1fVnw/M2gBM9cRbDZw/ED+p9eZbCdzeTgFuBI5KsnfrHHEUsLHN\nuzPJyratkwbWJUnqsZHuxTcGzwZ+DbgyyeUt9j+Bt9KNP3Uy8A3ghDbvAuBYYCvwXeAVAFW1I8mb\ngUtbuzdV1Y42/WrgLLoxrS5sD0lSz020QFXVpxh+nQjgyCHti27oj2HrWgesGxLfDDx1HmlKkiZg\n4tegJEkaxgIlSeolC5QkqZcsUJKkXrJASZJ6yQIlSeolC5QkqZcsUJKkXrJASZJ6yQIlSeolC5Qk\nqZcsUJKkXrJASZJ6yQIlSeolC5QkqZcsUJKkXrJASZJ6adJDvmuBLVv7oYlt+/q3vmhi25b00OMR\nlCSplyxQkqReskBJknrJAiVJ6iULlCSplyxQkqReskBJknppogUqyboktyT50kBsnySbkmxpz3u3\neJKcnmRrki8mOWxgmdWt/ZYkqwfiz0xyZVvm9CRZ3HcoSZqrSR9BnQWsmhZbC1xUVcuBi9prgGOA\n5e2xBjgDuoIGnAocARwOnDpV1FqbNQPLTd+WJKmnJlqgquoTwI5p4eOA9W16PXD8QPzs6lwC7JXk\nAOBoYFNV7aiqW4FNwKo2b8+q+kxVFXD2wLokST036SOoYfavqpsA2vN+LX4gcMNAu20ttrP4tiHx\noZKsSbI5yebt27fP+01IkuanjwVqJsOuH9Uc4kNV1ZlVtaKqVixdunSOKUqSFkofC9TN7fQc7fmW\nFt8GHDzQ7iDgxl3EDxoSlyQ9CPSxQG0ApnrirQbOH4if1HrzrQRub6cANwJHJdm7dY44CtjY5t2Z\nZGXrvXfSwLokST030eE2krwHeB6wb5JtdL3x3gq8L8nJwDeAE1rzC4Bjga3Ad4FXAFTVjiRvBi5t\n7d5UVVMdL15N11PwUcCF7SFJehCYaIGqqpfNMOvIIW0LOGWG9awD1g2JbwaeOp8cJUmT0cdTfJIk\nWaAkSf1kgZIk9ZIFSpLUSxYoSVIvWaAkSb1kgZIk9ZIFSpLUSxYoSVIvWaAkSb000Vsd6aFl2doP\nTXT717/1RRPdvqSF5RGUJKmXLFCSpF6yQEmSeskCJUnqJQuUJKmXLFCSpF6yQEmSeskCJUnqJQuU\nJKmXvJOEHjImeScL72IhLTyPoCRJvWSBkiT1kgVKktRLXoOSFoDXv6SF5xGUJKmXdosjqCSrgL8A\n9gD+tqreOuGUpAXjOFx6qHrIF6gkewB/DfwssA24NMmGqvryZDOTHho8valxecgXKOBwYGtVXQeQ\n5FzgOMACJT3ITfrocXe1WP8Y7A4F6kDghoHX24AjpjdKsgZY017+a5Jr57i9fYFvz3HZxdDn/Mxt\nbvqcG/Q7P3Obg7xt3rn9+Gwa7Q4FKkNi9YBA1ZnAmfPeWLK5qlbMdz3j0uf8zG1u+pwb9Ds/c5ub\nxcptd+jFtw04eOD1QcCNE8pFkjRLu0OBuhRYnuSQJA8HTgQ2TDgnSdIuPORP8VXV3UleA2yk62a+\nrqquGuMm532acMz6nJ+5zU2fc4N+52duc7MouaXqAZdjJEmauN3hFJ8k6UHIAiVJ6iUL1AJKsirJ\ntUm2Jlk7ge0fnOTiJFcnuSrJb7b4aUm+meTy9jh2YJk3tHyvTXL0mPO7PsmVLYfNLbZPkk1JtrTn\nvVs8SU5vuX0xyWFjzOtJA/vm8iR3JHntJPdbknVJbknypYHYyPsqyerWfkuS1WPM7U+SXNO2/4Ek\ne7X4siT/PrAP3zWwzDPbz8PWlv+wr4QsRG4jf47j+l2eIb/3DuR2fZLLW3yx991Mfz8m93NXVT4W\n4EHXAeOrwBOAhwNXAIcucg4HAIe16R8FvgIcCpwG/M6Q9oe2PB8BHNLy32OM+V0P7Dst9sfA2ja9\nFnhbmz4WuJDue2wrgc8u4uf4LbovEk5svwHPBQ4DvjTXfQXsA1zXnvdu03uPKbejgCVt+m0DuS0b\nbDdtPZ8DntXyvhA4Zky5jfQ5jvN3eVh+0+b/GfD7E9p3M/39mNjPnUdQC+feWypV1feBqVsqLZqq\nuqmqPt+m7wSupruTxkyOA86tqruq6mvAVrr3sZiOA9a36fXA8QPxs6tzCbBXkgMWIZ8jga9W1dd3\n0mbs+62qPgHsGLLdUfbV0cCmqtpRVbcCm4BV48itqj5SVXe3l5fQfd9wRi2/PavqM9X9VTt74P0s\naG47MdPnOLbf5Z3l146Cfgl4z87WMcZ9N9Pfj4n93FmgFs6wWyrtrDiMVZJlwDOAz7bQa9ph+Lqp\nQ3QWP+cCPpLksnS3lgLYv6pugu4XBNhvQrlNOZH7/4How36bMuq+mlSer6T7z3rKIUm+kOTjSZ7T\nYge2fBYrt1E+x0ntt+cAN1fVloHYRPbdtL8fE/u5s0AtnFndUmkxJPkR4P3Aa6vqDuAM4CeApwM3\n0Z1GgMXP+dlVdRhwDHBKkufupO2i7890X+R+MfAPLdSX/bYrM+UziX34RuBu4JwWugl4fFU9A/gt\n4O+T7LnIuY36OU7q830Z9//naCL7bsjfjxmbzpDHguVngVo4vbilUpKH0f1wnVNV/whQVTdX1Q+q\n6h7gf3Pf6ahFzbmqbmzPtwAfaHncPHXqrj3fMoncmmOAz1fVzS3PXuy3AaPuq0XNs10M/zngV9qp\nJ9rps++06cvoru38ZMtt8DTg2HKbw+e46J9vkiXAzwPvHch70ffdsL8fTPDnzgK1cCZ+S6V2Dvvd\nwNVV9ecD8cFrNy8BpnoQbQBOTPKIJIcAy+kuvo4jt0cn+dGpabqL6l9qOUz18lkNnD+Q20mtp9BK\n4Pap0wxjdL//YPuw36YZdV9tBI5Ksnc7rXVUiy24dIOCvh54cVV9dyC+NN2YbCR5At2+uq7ld2eS\nle3n9qSB97PQuY36OU7id/mFwDVVde+pu8XedzP9/WCSP3fz7fnh4369YI6l6/nyVeCNE9j+f6E7\nlP4icHl7HAv8HXBli28ADhhY5o0t32tZgJ5AO8ntCXS9oa4ArpraP8BjgYuALe15nxYP3UCTX225\nrxjzvvth4DvAYwZiE9tvdIXyJuA/6P4jPXku+4ruetDW9njFGHPbSnfdYern7l2t7S+0z/sK4PPA\nfxtYzwq6YvFV4K9od7YZQ24jf47j+l0ell+LnwW8alrbxd53M/39mNjPnbc6kiT1kqf4JEm9ZIGS\nJPWSBUqS1EsWKElSL1mgJEm9ZIGSdhNJKsnHJp2HNFsWKGmO0g2HUEnOmxY/q8UryUtnWHaPdENA\nTLX7kYF5z2uxPx33e5D6zAIljc/d3PcN/OmOAh7X2iyWp9DddUB6ULBASeMzdcuXYcOEvJzuzhWX\nLlYyVXVNVX1jsbYnzZcFShqfs+gGv/vVwWC60WZfTHfbm+/PdyNJnpzknCRfT3JXku1JLknyG9Pa\n3e8aVJKXD5xiHPb42LTl90rytiRfSfK9tp1zkzxxvu9BGmbJpBOQHsKuobv56GrgTwbiLwMeSTf4\n27yuMyU5iG7MniXAPwHfoBvJ9D8Bvwb85U4Wvxz4gyHxp9MNRne/m74Cn6K7m/YmuhuGHkh3v7if\nTfKsqvrKfN6LNJ0FShqvs4B3JnlmdUMmQHd676qq2tzdQHpefh7YEzi+qu53R+skj93ZglU1dUPQ\nwWUeR3fa8TbgdQOz/pLubtq/UPcNw0CSw4FPA++gu7GotGA8xSeN13uAu2idJZI8hW48orMWeDv/\nNj1QbSyh2UrySLqjsP2BE6vq2hZfCpwAbBgsTm0bn2vLHJ3kMXPMXRrKIyhpjKrqtiTnA7+c5Hfo\njp5+APzfBdrEB4E/Av4pyXuBjwCfqLmNnbUO+BngdVU1OH7PCrp/ZvdMctqQ5R7X5i8HNs9hu9JQ\nFihp/NYDv0TXMeJXgY1V9a2FWHFVfS3JfwbeRHdt65UAST4N/HZVfXY260nye235dVX1jmmz92nP\nz2+PmTx6lNylXfEUnzR+G+kGqXsH3dHGWQu58qq6oqqOoyskz2vb+RngwiT77mr5JC+hK3CfAl49\npMkd7fm0qspOHh9fkDckNRYoacyqauqU3oHArYxp+PCq+l5VfbyqXge8HdgbePbOlknyNLoRZ78B\n/HxVDev2vplupNWVC5yytFMWKGlx/BnwEuBFVXXXQq00yc/McJS0f3uecVtJ9uO+YnlcVW0f1q5d\nz3o/sCrJK4esZ0mSnRZCaS68BiUtgqq6ma6320L7FeBVSS4GtgL/Ttep4fnAFcBHd7LsacDjgU8C\nL2mn+gZdX1VntelXA4cC707y63Rd0e8Cfhx4Dt2R4ZMX4P1I97JASQ9u7wF+mO5U3rOBAF+n+wLu\nO2Y4ZTflh9vzc9pjuo/TrpdV1beTHEH33agT6Dpj/AD4JvAh4Jz5vhFpulTVpHOQJOkBvAYlSeol\nC5QkqZcsUJKkXrJASZJ6yQIlSeolC5QkqZcsUJKkXrJASZJ6yQIlSeql/w/umaAKb92aYgAAAABJ\nRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x103f22290>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "chrom_IMIs = []\n",
    "for chrom in chrlist:\n",
    "    subpos = ngenodf[ngenodf.Chrom==chrom]['Pos'].tolist()\n",
    "    chrom_IMIs.append([cryptofxns.IMI(subpos[i],subpos[i+1]) for i in range(len(subpos)-1)])        \n",
    "\n",
    "assert len(chrom_IMIs) == 14\n",
    "\n",
    "allIMIs = np.concatenate(chrom_IMIs)\n",
    "\n",
    "someIMIs = [imi for imi in allIMIs if imi <= 2000 ];\n",
    "print \" Total Number IMIs:\", len(allIMIs)\n",
    "print \"        IMIs < 2kb:\", len(someIMIs)\n",
    "print \"      % IMIs > 2kb:\", round(100 * (1 - (float(len(someIMIs))/len(allIMIs))),3);\n",
    "print \"   Median IMI size:\", np.median(allIMIs),\"bp\"\n",
    "\n",
    "plt.hist(someIMIs);\n",
    "plt.xlabel('IMI size',fontsize=19);\n",
    "plt.ylabel('Counts',fontsize=19);\n",
    "plt.tight_layout();\n",
    "plt.savefig('../FIGS/Crypto-IMI-Hist-Nov2017.png',dpi=900)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save the new dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#ngenodf.to_csv('../FILES/FiltGenotoHapFiltGeno%s'%fileendname,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
